File: node_second_line/src/profile_eval.mjs

export function evaluateRequest(profile, request) {
  if (!request || typeof request !== 'object' || Array.isArray(request)) {
    return { verdict: 'DENY', decisive_rule_ids: [], matched: [], failed: ['request_not_object'], fail_closed: true };
  }
  const obj = profile.mode === 'object' ? (request.object || request) : (request.after || (request.object || {}));
  if (!obj || typeof obj !== 'object' || Array.isArray(obj)) {
    return { verdict: 'DENY', decisive_rule_ids: [], matched: [], failed: ['object_missing'], fail_closed: true };
  }
  const matched = [];
  const failed = [];
  function metadataRuleId(node) {
    return node && node.metadata && typeof node.metadata.rule_id === 'string' ? node.metadata.rule_id : null;
  }
  function deepGet(data, p) {
    if (!p) return null;
    let cur = data;
    for (const part of p.split('.')) {
      if (!cur || typeof cur !== 'object' || Array.isArray(cur) || !(part in cur)) return null;
      cur = cur[part];
    }
    return cur;
  }
  function evalNode(node) {
    const kind = node.kind;
    const rid = metadataRuleId(node);
    let ok = false;
    if (kind === 'AND') ok = (node.children || []).every(evalNode);
    else if (kind === 'OR') ok = (node.children || []).some(evalNode);
    else if (kind === 'NOT') {
      const ch = node.children || [];
      ok = ch.length === 1 && !evalNode(ch[0]);
    } else if (kind === 'EQ') ok = deepGet(obj, node.path) === node.value;
    else if (kind === 'NEQ') ok = deepGet(obj, node.path) !== node.value;
    else if (kind === 'IN') ok = Array.isArray(node.value) && node.value.includes(deepGet(obj, node.path));
    else if (kind === 'NOT_IN') ok = Array.isArray(node.value) && !node.value.includes(deepGet(obj, node.path));
    else if (kind === 'RANGE') {
      const v = deepGet(obj, node.path);
      ok = typeof v === 'number' && v >= node.low && v <= node.high;
    } else if (kind === 'EXISTS') ok = deepGet(obj, node.path) !== null;
    else if (kind === 'ABSENT') ok = deepGet(obj, node.path) === null;
    else throw new Error(`unsupported node kind: ${kind}`);
    if (rid) (ok ? matched : failed).push(rid);
    return ok;
  }
  const ok = evalNode(profile.root);
  const decisive_rule_ids = ok ? [...matched] : [...failed];
  return { verdict: ok ? 'ALLOW' : 'DENY', decisive_rule_ids, matched, failed, fail_closed: false };
}

File: externalized_second_line/src/bcf_primitive_externalized/profile_eval.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class EvalResult:
    verdict: str
    decisive_rule_ids: list[str]
    matched: list[str]
    failed: list[str]
    fail_closed: bool


def _metadata_rule_id(node: dict[str, Any]) -> str | None:
    metadata = node.get('metadata')
    if isinstance(metadata, dict):
        rid = metadata.get('rule_id')
        return rid if isinstance(rid, str) else None
    return None


def _get_value(obj: dict[str, Any], path: str | None) -> Any:
    if path is None:
        return None
    cur: Any = obj
    for part in path.split('.'):
        if not isinstance(cur, dict) or part not in cur:
            return None
        cur = cur[part]
    return cur


def _eval_node(node: dict[str, Any], obj: dict[str, Any], matched: list[str], failed: list[str]) -> bool:
    kind = node['kind']
    rid = _metadata_rule_id(node)
    if kind == 'AND':
        ok = all(_eval_node(child, obj, matched, failed) for child in node.get('children', []))
    elif kind == 'OR':
        ok = any(_eval_node(child, obj, matched, failed) for child in node.get('children', []))
    elif kind == 'NOT':
        children = node.get('children', [])
        ok = len(children) == 1 and (not _eval_node(children[0], obj, matched, failed))
    elif kind == 'EQ':
        ok = _get_value(obj, node.get('path')) == node.get('value')
    elif kind == 'NEQ':
        ok = _get_value(obj, node.get('path')) != node.get('value')
    elif kind == 'IN':
        ok = _get_value(obj, node.get('path')) in list(node.get('value', []))
    elif kind == 'NOT_IN':
        ok = _get_value(obj, node.get('path')) not in list(node.get('value', []))
    elif kind == 'RANGE':
        value = _get_value(obj, node.get('path'))
        ok = isinstance(value, (int, float)) and value >= node.get('low') and value <= node.get('high')
    elif kind == 'EXISTS':
        ok = _get_value(obj, node.get('path')) is not None
    elif kind == 'ABSENT':
        ok = _get_value(obj, node.get('path')) is None
    else:
        raise ValueError(f'unsupported node kind: {kind}')
    if rid:
        (matched if ok else failed).append(rid)
    return ok


def evaluate_request(profile: dict[str, Any], request: dict[str, Any]) -> EvalResult:
    if not isinstance(request, dict):
        return EvalResult('DENY', [], [], ['request_not_object'], True)
    obj = request.get('object')
    if not isinstance(obj, dict):
        return EvalResult('DENY', [], [], ['object_missing'], True)
    matched: list[str] = []
    failed: list[str] = []
    root = profile['root']
    ok = _eval_node(root, obj, matched, failed)
    decisive = list(matched if ok else failed)
    return EvalResult('ALLOW' if ok else 'DENY', decisive, matched, failed, False)
